Bind crashed on macOS 15.4

### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
/Users/thueske [thueske@MBP-M3-Hueske] [21:54]
> brew config
HOMEBREW_VERSION: 4.4.26
ORIGIN: https://github.com/Homebrew/brew
HEAD: f1041842141f20bc701b522d4371d7843d905580
Last commit: 6 days ago
Branch: stable
Core tap JSON: 29 Mar 20:46 UTC
Core cask tap JSON: 29 Mar 20:46 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 11
Homebrew Ruby: 3.3.7 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.7/bin/ruby
CPU: 11-core 64-bit arm_lobos
Clang: 16.0.0 build 1600
Git: 2.39.5 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.4-arm64
CLT: 16.2.0.0.1.1733547573
Xcode: N/A
Rosetta 2: false



/Users/thueske [thueske@MBP-M3-Hueske] [21:53]
> brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed casks are deprecated or disabled.
You should find replacements for the following casks:
  svgcleaner


Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  python-psutil

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  icu4c@75
  icu4c@76
```

### Verification

- [x] My `brew doctor` output says `Your system is ready to brew.` and am still able to reproduce my issue.
- [x] I ran `brew update` and am still able to reproduce my issue.
- [x] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [x] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

Execute a `dig` command.

### What happened (include all command output)?

```
/Users/thueske [thueske@MBP-M3-Hueske] [21:53]
> dig -v
DiG 9.20.7
mem.c:367: INSIST(s >= size) failed, back trace
0   libisc-9.20.7.dylib                 0x0000000102873f3c default_callback + 72
1   libisc-9.20.7.dylib                 0x0000000102873ed4 isc_assertion_failed + 20
2   libisc-9.20.7.dylib                 0x0000000102887d30 isc__mem_strdup + 0
3   libxml2.2.dylib                     0x00000001a4f09764 xmlCleanupCharEncodingHandlers + 96
4   libxml2.2.dylib                     0x00000001a4f27cac xmlCleanupParser + 36
5   libisc-9.20.7.dylib                 0x000000010289a688 isc__xml_shutdown + 12
6   libisc-9.20.7.dylib                 0x00000001028818c0 isc__shutdown + 20
7   libsystem_c.dylib                   0x000000019b7c9944 __cxa_finalize_ranges + 480
8   libsystem_c.dylib                   0x000000019b7c9704 exit + 44
9   dig                                 0x00000001027ac224 dash_option + 2236
10  dig                                 0x00000001027aac24 parse_args + 9452
11  dig                                 0x00000001027a86d4 dig_query_setup + 64
12  dig                                 0x00000001027ab290 main + 44
13  dyld                                0x000000019b572b4c start + 6000
[1]    29393 abort      dig -v
```

### What did you expect to happen?

An output without any backtrace.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
Execute any `dig` command.
```

I'm able to reproduce this. Rebuilding this against Homebrew `libxml2` seems to help. The existing `bind` formula (using macOS-provided `libxml2`) also fails to build for me locally on macOS 15.4.

The build error seems to point to `libxml2` as well:
```
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
/bin/sh: line 1: 84738 Abort trap: 6           ./cfg_test --zonegrammar primary > primary.zoneopt
make[3]: *** [primary.zoneopt] Error 134
make[3]: *** Waiting for unfinished jobs....
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
/bin/sh: line 1: 84743 Abort trap: 6           ./cfg_test --zonegrammar stub > stub.zoneopt
make[3]: *** [stub.zoneopt] Error 134
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
/bin/sh: line 1: 84739 Abort trap: 6           ./cfg_test --zonegrammar secondary > secondary.zoneopt
make[3]: *** [secondary.zoneopt] Error 134
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
/bin/sh: line 1: 84746 Abort trap: 6           ./cfg_test --zonegrammar in-view > in-view.zoneopt
make[3]: *** [in-view.zoneopt] Error 134
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
/bin/sh: line 1: 84741 Abort trap: 6           ./cfg_test --zonegrammar hint > hint.zoneopt
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
make[3]: *** [hint.zoneopt] Error 134
/bin/sh: line 1: 84745 Abort trap: 6           ./cfg_test --zonegrammar redirect > redirect.zoneopt
/bin/sh: line 1: 84742 Abort trap: 6           ./cfg_test --zonegrammar forward > forward.zoneopt
make[3]: *** [redirect.zoneopt] Error 134
make[3]: *** [forward.zoneopt] Error 134
/bin/sh: line 1: 84740 Abort trap: 6           ./cfg_test --zonegrammar mirror > mirror.zoneopt
make[3]: *** [mirror.zoneopt] Error 134
xml.c:57:isc__xml_initialize(): fatal error: RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc, isc__xml_realloc, isc__xml_strdup) == 0) failed
/bin/sh: line 1: 84744 Abort trap: 6           ./cfg_test --zonegrammar static-stub > static-stub.zoneopt
make[3]: *** [static-stub.zoneopt] Error 134
make[2]: *** [all-recursive] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

CC @Bo98 and @cho-m who might know more about this.

--- a/Formula/b/bind.rb
+++ b/Formula/b/bind.rb
@@ -8,11 +8,21 @@ class Bind < Formula
   # "version_scheme" because someone upgraded to 9.15.0, and required a
   # downgrade.
 
-  url "https://downloads.isc.org/isc/bind9/9.20.7/bind-9.20.7.tar.xz"
-  sha256 "43323c8d22d2144282c37b4060ec11e98c24835e225688876fad08ba7b95dca6"
+  # TODO: Uncomment below when patch is no longer needed.
+  # url "https://downloads.isc.org/isc/bind9/9.20.7/bind-9.20.7.tar.xz"
+  # sha256 "43323c8d22d2144282c37b4060ec11e98c24835e225688876fad08ba7b95dca6"
   license "MPL-2.0"
+  revision 1
   version_scheme 1
   head "https://gitlab.isc.org/isc-projects/bind9.git", branch: "main"
+
+  # TODO: Remove `stable` block when patch is no longer needed.
+  stable do
+    url "https://downloads.isc.org/isc/bind9/9.20.7/bind-9.20.7.tar.xz"
+    sha256 "43323c8d22d2144282c37b4060ec11e98c24835e225688876fad08ba7b95dca6"
+
+    patch :DATA
+  end
 
   # BIND indicates stable releases with an even-numbered minor (e.g., x.2.x)
   # and the regex below only matches these versions.
@@ -104,3 +114,55 @@ class Bind < Formula
     system bin/"dig", "ü.cl"
   end
 end
+
+__END__
+diff --git i/lib/isc/xml.c w/lib/isc/xml.c
+index 7dd9424..af08a50 100644
+--- i/lib/isc/xml.c
++++ w/lib/isc/xml.c
+@@ -19,6 +19,7 @@
+ #include <libxml/parser.h>
+ #include <libxml/xmlversion.h>
+ 
++#ifndef __APPLE__
+ static isc_mem_t *isc__xml_mctx = NULL;
+ 
+ static void *
+@@ -44,17 +45,20 @@ isc__xml_free(void *ptr) {
+ 	isc_mem_free(isc__xml_mctx, ptr);
+ }
+ 
++#endif /* !__APPLE__ */
+ #endif /* HAVE_LIBXML2 */
+ 
+ void
+ isc__xml_initialize(void) {
+ #ifdef HAVE_LIBXML2
++#ifndef __APPLE__
+ 	isc_mem_create(&isc__xml_mctx);
+ 	isc_mem_setname(isc__xml_mctx, "libxml2");
+ 	isc_mem_setdestroycheck(isc__xml_mctx, false);
+ 
+ 	RUNTIME_CHECK(xmlMemSetup(isc__xml_free, isc__xml_malloc,
+ 				  isc__xml_realloc, isc__xml_strdup) == 0);
++#endif /* !__APPLE__ */
+ 
+ 	xmlInitParser();
+ #endif /* HAVE_LIBXML2 */
+@@ -64,13 +68,15 @@ void
+ isc__xml_shutdown(void) {
+ #ifdef HAVE_LIBXML2
+ 	xmlCleanupParser();
++#ifndef __APPLE__
+ 	isc_mem_destroy(&isc__xml_mctx);
++#endif /* !__APPLE__ */
+ #endif /* HAVE_LIBXML2 */
+ }
+ 
+ void
+ isc__xml_setdestroycheck(bool check) {
+-#if HAVE_LIBXML2
++#if defined(HAVE_LIBXML2) && !defined(__APPLE__)
+ 	isc_mem_setdestroycheck(isc__xml_mctx, check);
+ #else
+ 	UNUSED(check);

--- a/Formula/b/bind.rb
+++ b/Formula/b/bind.rb
@@ -32,13 +32,13 @@ class Bind < Formula
   end
 
   bottle do
-    sha256 arm64_sequoia: "475a19780d6d715095643e643e29234fc4a8957fdf5fe0771804b90ca2dd5c73"
-    sha256 arm64_sonoma:  "3543326f334022fe41d3ad73c307245b300ac1a1cfe95af00a58035923b4b3f3"
-    sha256 arm64_ventura: "cb80d80f5611f36f33133d685ec183a48157243a81ff5b753f5c95316605e9b2"
-    sha256 sonoma:        "58d1065019d4c9fa3aa6f1cdaa8dc9d0de38a584849319dc765dd2bb057db0f5"
-    sha256 ventura:       "9a4908c982aba2d52f16de4110a1d88fe00cc122694b655dfa02de3983b1d99d"
-    sha256 arm64_linux:   "0e003c29242ddc0b9fdb7180cb6266fb09ce4af72aac1aca0429a560e4c9d1f4"
-    sha256 x86_64_linux:  "eeab5b333e38bf01971fc4597bb4580cd131a78780d77e835a3f92f75abf2859"
+    sha256 arm64_sequoia: "e80ed22d86c4a5fef6c89e8bbfab65269a71032a6a52b6bb103a9685013d9ede"
+    sha256 arm64_sonoma:  "38364bf644834e6a254362c1b5e417ccb0830fc480dcfec917fe595b6bab7ff4"
+    sha256 arm64_ventura: "4c1dbe0b4db181ef597737f90bdee2ee232d5fecef83a1c2a0824d72ae0e60e0"
+    sha256 sonoma:        "ff24146494d636c33df4eb670bc0e28937234bb198653041fefa946daffe3a5d"
+    sha256 ventura:       "1f06f5ec73358e4b89c7b287fad3dfd816a3aea1d913a91c3b8e6bd46205555f"
+    sha256 arm64_linux:   "99756495e4d5931d4d5decbee54f21d256bec8129013fa1e5ce7cdb70db385ed"
+    sha256 x86_64_linux:  "9095b82ecd97dc33b68c941b38f5c18b55d598b4b9becb265ebab304d952263e"
   end
 
   depends_on "pkgconf" => :build

